### PR TITLE
Starting cloud integration changes to kubelet

### PR DESCRIPTION
**Pull request by addyess**
Configure kubernetes snaps in an environment where we're running on a known cloud

### CODE REVIEW

File: charms/kubernetes_snaps.py [hunk 1]
```diff
@@ -148,35 +149,22 @@ def configure_apiserver(
     if external_cloud_provider.has_xcp:
         api_opts["cloud-provider"] = "external"
 
+    # TODO: Check if cloud-config is still required
+    # api_cloud_config_path = cloud_config_path("kube-apiserver")
     if external_cloud_provider.name == "gce":
         # TODO: We need a charm for external-cloud-provider for GCP
         # based on https://github.com/kubernetes/cloud-provider-gcp/
+        api_opts["cloud-provider"] = "gce"
+        # api_opts["cloud-config"] = str(api_cloud_config_path)
```
Comment (addyess) [inline charms/kubernetes_snaps.py]: Sadly, we can't do this.  If we set that it loses the external provider.  We're gonna have to make an out-of-tree charm for GCE
```suggestion
```

File: charms/kubernetes_snaps.py [hunk 2]
```diff
@@ -211,10 +199,10 @@ def configure_apiserver(
 def configure_controller_manager(
     cluster_cidr,
     cluster_name,
+    external_cloud_provider: ExternalCloud,
     extra_args_config,
     kubeconfig,
     service_cidr,
-    external_cloud_provider: ExternalCloud,
```
Comment (addyess) [inline charms/kubernetes_snaps.py]: i'm ok if the arg order needs to change -- but we'd have to adjust the caller in kubernetes-control-plane

File: charms/kubernetes_snaps.py [hunk 3]
```diff
@@ -148,35 +149,22 @@ def configure_apiserver(
...
-            feature_gates.append("CSIMigrationGCE=false")
-    elif is_state("endpoint.vsphere.ready"):
-        if (1, 12) <= kube_version:
-            api_opts["cloud-provider"] = "vsphere"
-            api_opts["cloud-config"] = str(api_cloud_config_path)
-        if kube_version < (1, 26, 0):
-            feature_gates.append("CSIMigrationvSphere=false")
-    elif is_state("endpoint.azure.ready"):
+    elif external_cloud_provider.name == "azure":
         api_opts["cloud-provider"] = "azure"
```
Comment (addyess) [inline charms/kubernetes_snaps.py]: this would engage the in-tree azure provider.  We want to avoid this

File: charms/kubernetes_snaps.py [hunk 4]
```diff
@@ -239,15 +227,17 @@ def configure_controller_manager(
...
-    elif is_state("endpoint.gcp.ready"):
+    if external_cloud_provider.name == "gce":
+        # TODO: We need a charm for external-cloud-provider for GCP
+        # based on https://github.com/kubernetes/cloud-provider-gcp/
+        log.warning(
+            "GCP cloud-provider is currently available in-tree "
+            "but can be avoided configuring here if we use the "
+            "external tree provider"
+        )
         controller_opts["cloud-provider"] = "gce"
```
Comment (addyess) [inline charms/kubernetes_snaps.py]: nah, we can't use this -- it would engage the in-tree GCE provider

**Review by mateoflorido — APPROVED** [review]
🥇 🥇

**Pull request merged**